### PR TITLE
bazel: Make a number of small improvements

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR makes a number of small improvements to our Bazel setup that allows folks to start using it for running `environmentd` and testing with `sqllogictest`

* Adds a `.bazelversion` file so we can use [bazelisk](https://github.com/bazelbuild/bazelisk)
* Updates the `.bazelrc` to include `--explain` for all actions that support it, hopefully making future debugging easier
* Adds a `bin/bazel` command.
    * `bin/bazel gen` repins dependencies and generates `BUILD.bazel` files.
    * `bin/bazel <other>` is forwarded to `bazel`
* Adds a `--bazel` flag to `run.py` that so we can run `bin/environmentd` and `bin/sqllogictest` and build with Bazel
* Updates the `rust_toolchain` rules and includes hashes for artifact re-use
* Updates the Nightly CI step to use the new `bin/bazel`

### Motivation

Progress towards https://github.com/MaterializeInc/materialize/issues/26796
Fixes https://github.com/MaterializeInc/materialize/issues/27032

### Tips for reviewer

This PR is broken up into separate commits that can be reviewed separately

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/a

### CODE REVIEW

**Review by def- — APPROVED** [review]
CI changes lgtm

File: misc/python/materialize/cli/run.py [hunk 1]
```diff
@@ -158,41 +165,54 @@ def main() -> int:
...
+        else:
+            print("Disabled macOS Codesigning")
 
         if args.wrapper:
             command = shlex.split(args.wrapper)
         else:
             command = []
-        command.append(str(artifact_path / args.program))
+
+        mzdata = MZ_ROOT / "mzdata"
```
Comment (maddyblue) [inline misc/python/materialize/cli/run.py:194]: I'm not convinced this is a good place to put the binaries. mzdata historically only contained runtime data, and so was an easy directory to remove to reset an environment (this is perhaps not true today, but it once was). If that's still the case for mzdata (and thus folks may have like `rf -rf mzdata` somewhere in their local test scripts), somewhere else may be better.

**Review by maddyblue — APPROVED** [review]

File: misc/python/materialize/cli/run.py (hunk 1 above)
Comment (ParkMyCar) [reply misc/python/materialize/cli/run.py:194]: Totally makes sense! What do you think about introducing a `.gitignore`-d `build` directory at the top level of the repo and placing them there?

Comment (maddyblue) [reply misc/python/materialize/cli/run.py:194]: SGTM!

Comment (ParkMyCar) [reply misc/python/materialize/cli/run.py]: Ended up calling the new dir `mzbuild`, seemed to line up better with `mzdata`

**Pull request merged**